source/features/useful-not-found-page.tsx

```diff
@@ -107,17 +62,132 @@ function init(): false | void {
 	for (let i = bar.children.length - 2; i >= 0; i--) {
 		void checkAnchor(bar.children[i] as HTMLAnchorElement);
 	}
+}
+
+async function showHelpfulLinks(bar: Element): Promise<void> {
+	// What we should try to do to be helpful:
+	//  1. Check if the file is on the default branch
+	//  2. Check the current branch's history for the file (was it deleted or renamed)
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+	const urlToFileOnDefaultBranch = await (async () => {
+		const url = new GitHubURL(location.href);
+		const currentBranch = url.branch;
+		if (!currentBranch) {
+			return;
+		}
+
+		const defaultBranch = await getDefaultBranch();
+		if (currentBranch === defaultBranch) {
+			return;
+		}
+
+		url.assign({
+			branch: defaultBranch,
+		});
+		if (await is404(url.toString())) {
+			return;
+		}
+
+		return url.toString();
+	})();
+
+	if (urlToFileOnDefaultBranch) {
+		bar.after(
+			<p className="container mt-4 text-center">
+				View <a href={urlToFileOnDefaultBranch}>this {getType()}</a> on the default branch.
+			</p>,
+		);
+		return;
+	}
+
+	const change = await (async (): Promise<Record<string, any> | void> => {
+		const url = new GitHubURL(location.href);
+		const currentBranch = url.branch;
+		const {filePath} = url;
+		if (!currentBranch || !filePath) {
+			return;
+		}
+
+		const commits = await api.v3(`commits?path=${filePath}&sha=${currentBranch}&per_page=2`);
+		if (!commits[0]) {
+			return;
+		}
+
+		const commitInfo = await api.v3(`commits/${commits[0].sha as string}`);
+		const fileInfo = commitInfo.files.find((file: AnyObject) => [file.filename, file.previous_filename].includes(filePath));
+		if (!fileInfo) {
+			return;
+		}
+
+		const commitSha = commitInfo.sha.slice(0, 8);
+		const commitDate = commitInfo.commit.committer.date;
+		const linkToCommit = commitInfo.html_url;
+
+		url.assign({
+			route: 'commits',
+		});
+		const linkToCommitHistory = url.toString();
+
+		if (fileInfo.status === 'removed') {
+			return {
+				type: 'removed',
+				commitDetails: {
+					filePath,
+					commitSha, commitDate,
+					linkToCommit, linkToCommitHistory,
+					linkToLastVersion: fileInfo.blob_url,
+				},
+			};
+		}
+
+		if (fileInfo.status === 'renamed') {
+			url.assign({
+				route: 'blob',
+				branch: commitInfo.parents[0].sha,
+				filePath,
+			});
+
+			return {
+				type: 'renamed',
+				commitDetails: {
+					filePath: fileInfo.previous_filename, newFilePath: fileInfo.filename,
+					commitSha, commitDate,
+					linkToCommit, linkToCommitHistory,
```

Comment:
Don't use tricks just to save lines, that defeats the purpose. Lines of code are just a metric to quantify the amount of code, this just hides code from the metric.

Reply:
I think I got a bit carried away with the lines thing 😅 . Will change that